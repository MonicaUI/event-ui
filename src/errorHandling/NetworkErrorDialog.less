@import "../dialogs/styles/StandardDialog.less";
@import (reference) "~event-widgets/shared-styles/Variables.less";
@import (reference) "~event-widgets/shared-styles/Utility.less";
@import (reference) '~nucleus-core/less/VariablesColors.less';

.messageContainer {
  text-align: center;
  padding-top: 40px;
  overflow: hidden;
}

.messageIcon {
  font-size: 5.313em;
}

.dialogContainer {
  width: 700px;
  .dialogMediaQuery();
}

.errorBox {
  border: 1px solid @primaryGrey;
  border-radius: 0.094em;
  width: ~"calc(100% - 5em)";
  margin: 2em auto;
  padding: 1em;

  > div {
    margin: 1em;
  }
}

.errorBoxLabel {
  font-weight: bold;
}
