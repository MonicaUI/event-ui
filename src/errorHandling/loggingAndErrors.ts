import Logger, {
  DebugLevels,
  setDebugLevel,
  setLogClient,
  CompositeLogClient,
  ConsoleLogClient
} from '@cvent/nucleus-logging';
import RemoteLogClient from '@cvent/nucleus-remote-log-client';
import { SessionTimedOutError } from '../dialogs/SessionTimedOutDialog';
import { ServiceError } from '@cvent/event-ui-networking';
import redirectToErrorPage from '../utils/redirectToErrorPage';
import { httpLogPageLoadId } from '@cvent/nucleus-networking';
import { debounce } from 'lodash';

const LOG = new Logger('event-guestside-site/GLOBAL');

let loggingAppSettings;
let remoteClient;

// For tests
export function setupLoggingForTest(remoteClientValue: $TSFixMe, appSettings?: $TSFixMe): $TSFixMe {
  remoteClient = remoteClientValue;
  loggingAppSettings = appSettings;
}

const logDevelopmentError = debounce(
  exception => {
    if (exception.stack && exception.stack.indexOf('invokeGuardedCallbackDev') >= 0) {
      /*
       * Ignore errors that will be processed by componentDidCatch.
       * SEE: https://github.com/facebook/react/issues/10474
       * This should only matter with development build because errors are rethrown to facilitate debugging
       */
      return;
    }
    // eslint-disable-next-line no-alert
    alert(
      'A fatal error occurred. This will cause the instance page to appear in production. ' +
        'Look at console for more details on what occurred. To enable fatal error page ' +
        'run the command "IS_DEBUG=false pnpm dev"'
    );
  },
  30000,
  { leading: true }
);

export async function logErrorAndRedirect(
  message: $TSFixMe,
  exception: $TSFixMe,
  ...rest: $TSFixMe[]
): Promise<$TSFixMe> {
  if (exception instanceof SessionTimedOutError || isAbortError(exception)) {
    return;
  }

  const errorType = exception instanceof ServiceError ? 'ServiceError' : 'ClientError';

  LOG.error(message, errorType, exception, ...rest, { version: (window as $TSFixMe).version });
  if (remoteClient) {
    const flushPromise = remoteClient.flushNow();
    const timeoutPromise = new Promise(resolve => setTimeout(resolve, 2500));
    await Promise.race([flushPromise, timeoutPromise]);
  }

  try {
    if (loggingAppSettings.isDebug) {
      logDevelopmentError(exception);
    } else {
      redirectToErrorPage(
        loggingAppSettings.viewRoot,
        loggingAppSettings.eventContext.eventId,
        loggingAppSettings.environment,
        exception.httpLogRequestId
      );
    }
  } catch (e) {
    console.error('Error logging.', e); // eslint-disable-line no-console
  }
}

function isAbortError(error) {
  // Connection was aborted by something, we don't care about this
  return error.name === 'AbortError';
}

function isFetchError(error) {
  // Fetch network request completely failed, e.g. network offline
  return error.name === 'FetchError';
}

// Error when a webpack chunk fails to load
function isChunkLoadError(error) {
  // Webpack chunk failed to load
  return error.name === 'ChunkLoadError';
}

export function isNetworkError(error: $TSFixMe): $TSFixMe {
  return isChunkLoadError(error) || isFetchError(error);
}

function isThirdPartyError(error) {
  /*
   * Lack of error information usually indicates error in a script from another domain that we are blocked from seeing
   * much information about because of cross-origin restrictions. We aren't responsible for these errors (most likely
   * some stupid tracking script the client is using), so don't do anything about them.
   */
  return !error || (!error.line && !error.column && !error.stack);
}

const assetDomain = process.env.NODE_ENV === 'production' ? 'cvent-assets.com' : 'localhost';

function isProbablyCodeSnippetOrExtension(error, { file }: $TSFixMe = {}) {
  // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
  if (file && file.includes(assetDomain)) {
    return false;
  }
  // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
  return !(error && error.stack && error.stack.includes(assetDomain));
}

let handleNetworkError;

export function runNetworkErrorHandler(): $TSFixMe {
  return handleNetworkError();
}

export function setNetworkErrorHandler(handler: $TSFixMe): $TSFixMe {
  handleNetworkError = handler;
}

export function initializeLogging(appSettings: $TSFixMe, eventId: $TSFixMe): $TSFixMe {
  const { debugLevel = DebugLevels.WARN, logToServer = true, environment, serviceBaseUrl } = appSettings;
  loggingAppSettings = appSettings;

  setDebugLevel(debugLevel);
  setNetworkErrorHandler(error => logErrorAndRedirect('Global Unhandled Network Error During Page Load', error));
  if (logToServer) {
    remoteClient = new RemoteLogClient({
      // @ts-expect-error ts-migrate(2322) FIXME: Type '{ httpLogPageLoadId: string; }' is not assig... Remove this comment to see the full error message
      context: { httpLogPageLoadId },
      url: `${serviceBaseUrl}/v1/log?environment=${environment}&eventId=${eventId}`,
      maxMessagesPerInterval: 200,
      maxWaitMs: 5000
    });
    const compositeClient = new CompositeLogClient(remoteClient, new ConsoleLogClient());
    setLogClient(compositeClient);
  }
  global.onunhandledrejection = async event => {
    let reason;
    try {
      reason = await event.reason;
      if (!reason) {
        // Probably an error in a third-party script because we don't manually use reject
        LOG.warn({ message: 'unhandledrejection event had no reason', event });
        return;
      }
    } catch (error) {
      reason = error;
      if (!reason) {
        reason = { message: 'unhandledrejection event reason promise rejected with no value', event };
      }
    }
    if (isNetworkError(reason)) {
      LOG.info('network error happened', reason);
      handleNetworkError(reason);
    } else if (isThirdPartyError(reason)) {
      LOG.warn('Third Party Script Unhandled Rejection', reason);
    } else if (isProbablyCodeSnippetOrExtension(reason)) {
      LOG.warn('Unhandled Promise Probably Code Snippet or Extension', reason);
    } else {
      void logErrorAndRedirect('Global Unhandled Promise Rejection', reason);
    }
  };
  window.onerror = (message, file, line, column, error) => {
    if (isThirdPartyError(error)) {
      /*
       * Error in a script that we are prevented from seeing because of cross-origin party
       * Probably a third party script in a code snippet that we aren't responsible for
       */
      LOG.warn('Third Party Script Unhandled Error', error, { message, file, line, column });
    } else if (isProbablyCodeSnippetOrExtension(error, { message, file, line, column })) {
      LOG.warn('Unhandled Error Probably Code Snippet or Extension', error, { message, file, line, column });
    } else {
      void logErrorAndRedirect('Global Unhandled Error', error, { message, file, line, column });
    }
  };
}
