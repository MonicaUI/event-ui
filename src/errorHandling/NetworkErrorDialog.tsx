import React from 'react';
import StandardDialog from '../dialogs/shared/StandardDialog';
import { openDialogContainer, closeDialogContainer } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import DialogStyles from './NetworkErrorDialog.less';
import { setNetworkErrorHandler } from './loggingAndErrors';
import getDialogContainerStyle from '../dialogs/shared/getDialogContainerStyle';
import { merge } from 'lodash';
import ThemeableComponent from 'nucleus-widgets/lib/ThemeableComponent';
import { resolve } from '@cvent/nucleus-dynamic-css';
import { httpLogPageLoadId } from '@cvent/nucleus-networking';

/**
 * Function to get the style object to pass into the event status dialog.
 */
function dialogStyle(state) {
  const { global, sections } = state.website.theme;
  return {
    ...global,
    dialogHeader: { styleMapping: 'header3' },
    subMessage: { styleMapping: 'body1' },
    errorBox: { styleMapping: 'body1' },
    dragHandle: { ...merge({}, global, sections.content2), styleMapping: 'custom' }
  };
}

class NetworkErrorDialog extends ThemeableComponent<$TSFixMe, $TSFixMe> {
  getElementBackground: $TSFixMe;
  getElementInlineStyle: $TSFixMe;
  getElementText: $TSFixMe;
  props: $TSFixMe;
  getStyleObject() {
    return {
      dialogHeader: this.getElementInlineStyle('dialogHeader'),
      subMessage: this.getElementInlineStyle('subMessage'),
      errorBox: this.getElementText('errorBox'),
      dragHandle: this.getElementBackground('dragHandle')
    };
  }
  render() {
    const { translate, closeDialog, date } = this.props;
    const style = this.getStyleObject();
    return (
      <StandardDialog
        onClose={closeDialog}
        closeFallbackText={translate('EventGuestSide_StandardDialog_FallbackText_Close__resx')}
        title={translate('EventGuestSide_NetworkError_Title__resx')}
        subMessage={translate('EventGuestSide_NetworkError_CheckYourConnection__resx')}
        icon="error"
        iconModifier="error"
        classes={DialogStyles}
        style={style}
      >
        <div {...resolve({ classes: DialogStyles, style }, 'errorBox')}>
          <div>{translate('EventGuestSide_NetworkError_ContactCustomerCenter__resx')}</div>
          <div>
            <span className={DialogStyles.errorBoxLabel}>{translate('Flex_ErrorPage_Time_Label__resx')}</span>:{' '}
            {date.toISOString()}
          </div>
          <div>
            <span className={DialogStyles.errorBoxLabel}>{translate('Flex_ErrorPage_sessionId_Label__resx')}</span>:{' '}
            {httpLogPageLoadId}
          </div>
        </div>
      </StandardDialog>
    );
  }
}

function openNetworkErrorDialog() {
  return (dispatch, getState) => {
    const { translate } = getState().text;
    const boundCloseDialog = () => {
      dispatch(closeDialogContainer());
    };
    const dialog = (
      <NetworkErrorDialog
        closeDialog={boundCloseDialog}
        translate={translate}
        style={dialogStyle(getState())}
        date={new Date(Date.now())}
      />
    );
    return dispatch(
      openDialogContainer(dialog, boundCloseDialog, {
        classes: { dialogContainer: DialogStyles.dialogContainer },
        style: { dragContainer: getDialogContainerStyle(getState()) }
      })
    );
  };
}

export function initializeNetworkErrorDialog(store: $TSFixMe): $TSFixMe {
  setNetworkErrorHandler(() => store.dispatch(openNetworkErrorDialog()));
}
