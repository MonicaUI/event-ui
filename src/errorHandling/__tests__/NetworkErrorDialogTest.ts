/* eslint-env jest */
import { create } from 'react-test-renderer';
import { runNetworkErrorHandler } from '../loggingAndErrors';
import { initializeNetworkErrorDialog } from '../NetworkErrorDialog';
import { openDialogContainer } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import EventSnapshot from '../../../fixtures/EventSnapshot.json';

jest.mock('nucleus-guestside-site/src/redux/modules/dialogContainer');
jest.mock('@cvent/nucleus-networking', () => {
  return {
    httpLogPageLoadId: 'fake-page-load-id'
  };
});

Date.now = () => new Date('2019-09-10T00:00:00.000Z').valueOf();

const store = {
  getState() {
    return {
      website: {
        theme: EventSnapshot.eventSnapshot.siteEditor.website.theme
      },
      text: {
        translate: resx => resx
      }
    };
  },
  dispatch(action) {
    if (typeof action === 'function') {
      return store.dispatch(action(store.dispatch, store.getState));
    }
  }
};

test('NetworkErrorDialog matches snapshot', () => {
  initializeNetworkErrorDialog(store);

  runNetworkErrorHandler();

  expect(openDialogContainer).toHaveBeenCalled();
  const dialog = openDialogContainer.mock.calls[0][0];
  const component = create(dialog);
  expect(component).toMatchSnapshot();
});
