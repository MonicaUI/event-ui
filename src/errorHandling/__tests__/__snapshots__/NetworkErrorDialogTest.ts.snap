// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkErrorDialog matches snapshot 1`] = `
<div>
  <div
    aria-describedby="dialogHeader"
    className=""
    style={
      Object {
        "backgroundColor": "#DDE2E6",
        "borderRadius": 0,
        "hover": "css-87ndg2",
      }
    }
  >
    <div
      className=""
      id="dialogHeader"
      style={
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 0,
          "color": "#041532",
          "fontFamily": "PT Sans, sans-serif",
          "fontSize": "1.875rem",
          "fontStyle": "normal",
          "fontWeight": 400,
          "hover": "css-87ndg2",
          "lineHeight": "1.3",
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        }
      }
    >
      <h2
        className=""
        style={Object {}}
      >
        <div
          className=""
          data-cvent-id="close"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          style={Object {}}
          tabIndex="0"
        >
          <span
            className="container"
          >
            <span
              aria-label="EventGuestSide_StandardDialog_FallbackText_Close__resx"
              className=""
              role="img"
            />
          </span>
        </div>
        EventGuestSide_NetworkError_Title__resx
      </h2>
    </div>
  </div>
  <div
    className=""
    name="messageContainer"
    style={Object {}}
    tabIndex="-1"
  >
    <span
      className="container"
    >
      <span
        aria-hidden="true"
        className=""
        role="img"
      />
    </span>
    <p
      className=""
      style={
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 0,
          "color": "#273F69",
          "fontFamily": "PT Sans, sans-serif",
          "fontSize": 16,
          "fontStyle": "normal",
          "fontWeight": 400,
          "hover": "css-87ndg2",
          "lineHeight": "1.3",
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        }
      }
    >
      EventGuestSide_NetworkError_CheckYourConnection__resx
    </p>
    <div
      className=""
      style={
        Object {
          "color": "#273F69",
          "fontFamily": "PT Sans, sans-serif",
          "fontSize": 16,
          "fontStyle": "normal",
          "fontWeight": 400,
          "hover": "css-87ndg2",
          "lineHeight": "1.3",
        }
      }
    >
      <div>
        EventGuestSide_NetworkError_ContactCustomerCenter__resx
      </div>
      <div>
        <span
          className="errorBoxLabel"
        >
          Flex_ErrorPage_Time_Label__resx
        </span>
        :
         
        2019-09-10T00:00:00.000Z
      </div>
      <div>
        <span
          className="errorBoxLabel"
        >
          Flex_ErrorPage_sessionId_Label__resx
        </span>
        :
         
        fake-page-load-id
      </div>
    </div>
  </div>
</div>
`;
